source/slang/slang-ir-util.cpp

```diff
@@ -688,9 +693,45 @@ bool areCallArgumentsSideEffectFree(IRCall* call)
                     // are not dependent on whatever we do in the call here.
                     continue;
                 default:
-                    // Skip the call itself, since we are checking if the call has side effect.
+                    // Skip the call itself if the var is used as an argument to an out parameter
+                    // since we are checking if the call has side effect.
+                    // We can't treat the call as side effect free if var is used as an inout parameter,
+                    // because if the call is inside a loop there will be a visible side effect after
+                    // the call.
                     if (use->getUser() == call)
+                    {
+                        auto funcType = as<IRFuncType>(call->getCallee()->getDataType());
+                        if (!funcType)
+                            return false;
+                        if (funcType->getParamCount() > i && as<IROutType>(funcType->getParamType(i)))
+                            continue;
+
+                        // We are an argument to an inout parameter.
+                        // We can only treat the call as side effect free if the call is not inside a loop.
+                        // 
+                        // If we don't have the loop information here, we will conservatively return false.
+                        //
+                        if (!dom)
+                            return false;
+
+                        // If we have dominator tree available, use it to check if the call is inside a loop.
+                        auto callBlock = as<IRBlock>(call->getParent());
+                        if (!callBlock) return false;
+                        auto varBlock = as<IRBlock>(arg->getParent());
+                        if (!varBlock) return false;
+                        while (callBlock != varBlock)
+                        {
+                            callBlock = dom->getImmediateDominator(callBlock);
+                            if (!callBlock)
+                                return false; // If we are here, var does not dominate the call, which should never happen.
+                            if (as<IRLoop>(callBlock->getTerminator()))
+                                return false; // The var is used in a loop, must return false.
+                        }
+                        // If we reach here, the var is used as an inout parameter for the call, but the call
+                        // is not nested in a loop at an higher nesting level than where the var is defined,
+                        // so we can treat the use as side effect free.
```

Comment:
I like this analysis, although it starts to feel like our definition of "side effect free" is getting a bit loose. What we really mean here is DCE-able and/or CSE-able instructions, right?

Reply:
Done.